// @sanity/visual-editing/dist/index.js
import { useDocuments as useDocuments$1 } from "./_chunks-es/mutations.js";
import { getArrayItemKeyAndParentPath } from "./_chunks-es/mutations.js";
import { SharedStateContext, useOptimistic as useOptimistic$1 } from "./_chunks-es/SharedStateContext.js";
import { createOverlayController } from "./_chunks-es/SharedStateContext.js";
import { enableVisualEditing } from "./_chunks-es/enableVisualEditing.js";
import { c } from "react-compiler-runtime";
import { useContext, useSyncExternalStore } from "react";
import { createDataAttribute } from "@sanity/visual-editing-csm";
function useSharedState(key) {
  const $ = c(3), context = useContext(SharedStateContext);
  if (!context)
    throw new Error("useSharedState must be used within a SharedStateProvider");
  const {
    store
  } = context;
  let t0;
  return $[0] !== key || $[1] !== store ? (t0 = () => store.getState()[key], $[0] = key, $[1] = store, $[2] = t0) : t0 = $[2], useSyncExternalStore(store.subscribe, t0);
}
const useDocuments = useDocuments$1, useOptimistic = useOptimistic$1;
export {
  createDataAttribute,
  createOverlayController,
  enableVisualEditing,
  getArrayItemKeyAndParentPath,
  useDocuments,
  useOptimistic,
  useSharedState
};
