// @sanity/visual-editing/dist/index.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: !0 });
var mutations = require("./_chunks-cjs/mutations.cjs"), SharedStateContext = require("./_chunks-cjs/SharedStateContext.cjs"), enableVisualEditing = require("./_chunks-cjs/enableVisualEditing.cjs"), reactCompilerRuntime = require("react-compiler-runtime"), react = require("react"), visualEditingCsm = require("@sanity/visual-editing-csm");
function useSharedState(key) {
  const $ = reactCompilerRuntime.c(3), context = react.useContext(SharedStateContext.SharedStateContext);
  if (!context)
    throw new Error("useSharedState must be used within a SharedStateProvider");
  const {
    store
  } = context;
  let t0;
  return $[0] !== key || $[1] !== store ? (t0 = () => store.getState()[key], $[0] = key, $[1] = store, $[2] = t0) : t0 = $[2], react.useSyncExternalStore(store.subscribe, t0);
}
const useDocuments = mutations.useDocuments, useOptimistic = SharedStateContext.useOptimistic;
exports.getArrayItemKeyAndParentPath = mutations.getArrayItemKeyAndParentPath;
exports.createOverlayController = SharedStateContext.createOverlayController;
exports.enableVisualEditing = enableVisualEditing.enableVisualEditing;
Object.defineProperty(exports, "createDataAttribute", {
  enumerable: !0,
  get: function() {
    return visualEditingCsm.createDataAttribute;
  }
});
exports.useDocuments = useDocuments;
exports.useOptimistic = useOptimistic;
exports.useSharedState = useSharedState;
